### PR TITLE
Migrate collection item and argument scopes

### PR DESCRIPTION
**Pull request by AndreasArvidsson**
Note that we are here just migrating markdown and yaml collection. For languages like typescript I propose that the insertion delimiter is either `,  ` for an inline collection or `,\n` for an vertical collection. When the edit is constructed by our destination `\n` gets replaced by `{indent}\n` or `\n{indent}` depending if it's before or after. In short I don't think we actually should use leading or trailing for the insertion delimiters and the destination can take care of that on its own.

To get the correct insertion delimiter we could have a conditional insertion delimiter predicate that actually uses the collection and not the item itself to determine if it's a vertical collection/insertion delimiter.

`insertionDelimiter = @list.range.isSingleLine ? ", " : ",\n"`
```
(
  (array
    (_) @collectionItem
  ) @list
  (#conditional-insertion-delimiter! @collectionItem @list ", " ",\n")
 ) 
```

Edit: Just went ahead and migrated typescript argument using the above predicate. Personally I think it turned out quite nicely.

Will fix the second task
https://github.com/cursorless-dev/cursorless/issues/585

## Checklist

- [x] I have added [tests](https://www.cursorless.org/docs/contributing/test-case-recorder/)
- [-] I have updated the [docs](https://github.com/cursorless-dev/cursorless/tree/main/docs) and [cheatsheet](https://github.com/cursorless-dev/cursorless/tree/main/cursorless-talon/src/cheatsheet)
- [-] I have not broken the cheatsheet

### CODE REVIEW

File: packages/cursorless-engine/src/languages/TreeSitterQuery/QueryCapture.ts [hunk 1]
```diff
@@ -38,6 +38,9 @@ export interface QueryCapture {
 
   /** The insertion delimiter to use if any */
   readonly insertionDelimiter: string | undefined;
+
+  /** The insertion prefix to use if any */
+  readonly insertionPrefix: string | undefined;
```
Comment (AndreasArvidsson) [inline packages/cursorless-engine/src/languages/TreeSitterQuery/QueryCapture.ts:43]: Separate the insertion delimiter that is used to separate each item in a collection from the insertion prefix which only goes before the text content and that is used on every item even if the collection only has one. Hence this is not actually a delimiter.

File: packages/cursorless-engine/src/languages/TreeSitterQuery/queryPredicateOperators.ts [hunk 2]
```diff
@@ -214,6 +214,20 @@ class InsertionDelimiter extends QueryPredicateOperator<InsertionDelimiter> {
   }
 }
 
+class InsertionPrefix extends QueryPredicateOperator<InsertionPrefix> {
+  name = "insertion-prefix!" as const;
+  schema = z.union([z.tuple([q.node, q.string]), z.tuple([q.node, q.node])]);
+
+  run(nodeInfo: MutableQueryCapture, prefix: string | MutableQueryCapture) {
```
Comment (AndreasArvidsson) [inline packages/cursorless-engine/src/languages/TreeSitterQuery/queryPredicateOperators.ts:221]: Support either a literal prefix or a node to extract it from

File: packages/cursorless-vscode-e2e/src/suite/fixtures/recorded/languages/markdown/chuckItem2.yml [hunk 3]
```diff
@@ -21,8 +21,7 @@ initialState:
 finalState:
   documentContents: |-
     - aaa
```
Comment (AndreasArvidsson) [inline packages/cursorless-vscode-e2e/src/suite/fixtures/recorded/languages/markdown/chuckItem2.yml:23]: We don't leave an empty line anymore

File: queries/markdown.scm
Comment (pokey) [inline queries/markdown.scm:1]: test for "clone" / "pour"

**Comment by AndreasArvidsson** [comment]
Another thing that just struck me(the item patterns are getting quite crowded) is why do we need to give a start and an end to the leading and trailing ranges? shouldn't they always terminate at the content range(or extended content range)? I think we can just use `@_.leading` and `@_.trailing` with optional `.startOf`/`.endOf`. We already know that the leading terminates at the start of the content range and the trailing starts at the end of the content range.

File: queries/javascript.core.scm [hunk 4]
```diff
@@ -664,3 +664,36 @@
...
+(
+  (formal_parameters
+    (_)? @_.leading.start.endOf
+    .
+    (_) @argumentOrParameter @_.leading.end.startOf @_.trailing.start.endOf
+    .
+    (_)? @_.trailing.end.startOf
+  ) @dummy
+  (#not-type? @argumentOrParameter "comment")
+  (#conditional-insertion-delimiter! @argumentOrParameter @dummy ", " ",\n")
```
[the remark this answers is not in the document]
Comment (josharian) [reply queries/javascript.core.scm:679]: `delimiter-ternary`?

[the remark this answers is not in the document]
Comment (pokey) [reply queries/javascript.core.scm:679]: Hmm would prefer if it captured the fact that it has to do with whether it's multiline

File: packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/TreeSitterScopeHandler/TreeSitterScopeHandler.ts [hunk 5]
```diff
@@ -70,6 +70,17 @@ export class TreeSitterScopeHandler extends BaseTreeSitterScopeHandler {
...
 
+    const rawPrefixRange = getRelatedRange(
+      match,
+      scopeTypeType,
+      "prefix",
+      true,
+    );
+    const prefixRange =
+      rawPrefixRange != null
+        ? new Range(rawPrefixRange.start, contentRange.start)
```
[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/TreeSitterScopeHandler/TreeSitterScopeHandler.ts:81]: I really don't like having to do unnecessary syntax, but sure we can await that discussion.

**Review by pokey — APPROVED** [review]
🎉

File: packages/cursorless-engine/src/processTargets/modifiers/EveryScopeStage.ts [hunk 6]
```diff
@@ -73,13 +73,17 @@ export class EveryScopeStage implements ModifierStage {
...
       // If target had no explicit range, or was contained by a single target
       // instance, expand to iteration scope before overlapping
-      scopes = this.getDefaultIterationRange(
-        scopeHandler,
-        this.scopeHandlerFactory,
-        target,
-      ).flatMap((iterationRange) =>
-        getScopesOverlappingRange(scopeHandler, editor, iterationRange),
-      );
+      try {
```
[the remark this answers is not in the document]
Comment (phillco) [reply packages/cursorless-engine/src/processTargets/modifiers/EveryScopeStage.ts:76]: Just commenting from the future to say that we removed this :D https://github.com/cursorless-dev/cursorless/pull/2740

**Pull request merged**